src/ggml-rpc.cpp

```diff
@@ -0,0 +1,622 @@
+        (void*)tensor->data, request->offset(), request->data().size());
+    ggml_backend_tensor_set(tensor, request->data().c_str(), request->offset(), request->data().size());
+    ggml_free(ctx);
+    return grpc::Status::OK;
+}
+
+grpc::Status BackendImpl::GetTensor(grpc::ServerContext* context, const ggml::GetTensorRequest* request, ggml::GetTensorReply* reply) {
+    UNUSED(context);
+    struct ggml_init_params params {
+            /*.mem_size   =*/ ggml_tensor_overhead(),
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ true,
+    };
+    struct ggml_context * ctx = ggml_init(params);
+    struct ggml_tensor * tensor = deserialize_tensor(ctx, request->tensor());
+    GGML_PRINT_DEBUG("GetTensor, bufptr: %p, dataptr: %p, offset: %u, size: %u\n", (void*)tensor->buffer,
+        (void*)tensor->data, request->offset(), request->size());
+    std::string data(request->size(), 0);
+    ggml_backend_tensor_get(tensor, &data[0], request->offset(), request->size());
+    reply->set_data(data);
+    ggml_free(ctx);
+    return grpc::Status::OK;
+}
+
+grpc::Status BackendImpl::CopyTensor(grpc::ServerContext* context, const ggml::CopyTensorRequest* request, ggml::CopyTensorReply* reply) {
+    UNUSED(context);
+    struct ggml_init_params params {
+            /*.mem_size   =*/ 2*ggml_tensor_overhead(),
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ true,
+    };
+    struct ggml_context * ctx = ggml_init(params);
+    struct ggml_tensor * src = deserialize_tensor(ctx, request->src());
+    struct ggml_tensor * dst = deserialize_tensor(ctx, request->dst());
+    GGML_PRINT_DEBUG("CopyTensor, src: %p, dst: %p\n", (void*)src, (void*)dst);
+    bool result = ggml_backend_buffer_copy_tensor(src, dst);
+    reply->set_result(result);
+    ggml_free(ctx);
+    return grpc::Status::OK;
+}
+
+static struct ggml_tensor * create_node(uint64_t id,
+                                        struct ggml_context * ctx,
+                                        const ggml::GraphComputeRequest* request,
+                                        std::unordered_map<uint64_t, struct ggml_tensor*> & tensor_map) {
+    if (id == 0) {
+        return nullptr;
+    }
+    if (tensor_map.find(id) != tensor_map.end()) {
+        return tensor_map[id];
+    }
+    for (int i = 0; i < request->tensors_size(); i++) {
+        if (request->tensors(i).id() == id) {
+            GGML_PRINT_DEBUG("create node: %lx\n", id);
+            const ggml::Tensor & protobuf_tensor = request->tensors(i);
+            struct ggml_tensor * result = deserialize_tensor(ctx, protobuf_tensor);
+            for (int i = 0; i < GGML_MAX_SRC; i++) {
+                result->src[i] = create_node(protobuf_tensor.src(i), ctx, request, tensor_map);
+            }
+            result->view_src = create_node(protobuf_tensor.view_src(), ctx, request, tensor_map);
+            result->view_offs = protobuf_tensor.view_offs();
+            tensor_map[id] = result;
```

Comment:
This is probably not the issue, but the tensor should be added to the map before calling `create_node`, so that it is not created again if there is a reference to itself in `src`, (which can happen in `ggml_cast`).

Reply:
Nice catch. Still doesn't fix gpt-2 though